File: Dynamic Programming/GFG/Knapsack Memoization.cpp
#include<bits/stdc++.h>
using namespace std;
int memorize_dp(int wt[],int val[],int W,int n,vector<vector<int>>K)
{
    
    //base conditon 
    if(n==0 || W==0)
    {
        return 0;
    }
    
    if(K[n][W]!=-1)
    {
        return K[n][W];
    }
    
    
    if(wt[n-1]>W)
    {
        // in this case we can't put it in the bag
        K[n][W]= memorize_dp(wt,val,W,n-1,K);
        return K[n][W];
        
    }
    else 
    {
        // int this case you have recursively call and choose whether in last element if we recursively call and add value to it or not
        K[n][W]= max(val[n-1]+memorize_dp(wt,val,W-wt[n-1],n-1,K),memorize_dp(wt,val,W,n-1,K));
        return K[n][W];
    }
}

//also do by defining another function 
// int knapsack(int wt[],int val[],int W,int n)
// {
//      vector<vector<int>> K(n + 1, vector<int>(W + 1,-1));
//      return memorize_dp(wt,val,W,n,K);
     
// }

int main()
{
    int wt[] = { 10, 20, 30 };
    int val[] = { 60, 100, 120 };
    int W = 50;
    int n = sizeof(val) / sizeof(val[0]);
    vector<vector<int>> K(n + 1, vector<int>(W + 1,-1));
    
    cout << memorize_dp(wt,val,W, n,K);
    
}

File: Dynamic Programming/GFG/Knapsack Memoization.py
val = [60, 100, 120 ]
wt = [10, 20, 30 ]
W=50
n=len(val)
#intialize 2d dp array with all value with -1
dp=[[-1 for i in range(W+1)] for j in range(n+1)]

def knapsack(val,wt,W,n):
    if n==0 or W==0:
        return 0
    if dp[n][W]!=-1:
        return dp[n][W]
    if wt[n-1]<=W:
        dp[n][W]=max(val[n-1]+knapsack(val,wt,W-wt[n-1],n-1),knapsack(val,wt,W,n-1))
        return dp[n][W]
    elif wt[n-1]>W:
        dp[n][W]=knapsack(val,wt,W,n)
        return dp[n][W]
print(knapsack(val,wt,W,n))
